java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -60,112 +61,165 @@
+     * conditions of switch statement are broader and not evaluated using this method.
+     *
+     * this method evaluates the given condition. if the result is either true or false,
+     * then the execution state is returned.
+     * otherwise, if the result is unknown, then the execution state is split in two branches
+     * with result TRUE and FALSE. whenever possible constraints are also set on the variables
+     * found in the condition.
+     *
+     * @param state execution state
+     * @param tree tree representing the condition
+     * @return output states
+     */
+    private List<ExecutionState> evaluateCondition(ExecutionState executionState, ExpressionTree tree) {
+      List<ExecutionState> result = new ArrayList<>();
+      ExpressionTree expressionTree = removeCastAndParenthesis(tree);
+      if (expressionTree.is(Tree.Kind.CONDITIONAL_AND)) {
+        evaluateConditionConditionalAnd(result, executionState, (BinaryExpressionTree) expressionTree);
+      } else if (expressionTree.is(Tree.Kind.CONDITIONAL_OR)) {
+        evaluateConditionConditionalOr(result, executionState, (BinaryExpressionTree) expressionTree);
+      } else if (expressionTree.is(Tree.Kind.EQUAL_TO)) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, SymbolicRelation.EQUAL_TO);
+      } else if (expressionTree.is(Tree.Kind.GREATER_THAN)) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, SymbolicRelation.GREATER_THAN);
+      } else if (expressionTree.is(Tree.Kind.GREATER_THAN_OR_EQUAL_TO)) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, SymbolicRelation.GREATER_EQUAL);
+      } else if (expressionTree.is(Tree.Kind.LESS_THAN)) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, SymbolicRelation.LESS_THAN);
+      } else if (expressionTree.is(Tree.Kind.LESS_THAN_OR_EQUAL_TO)) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, SymbolicRelation.LESS_EQUAL);
+      } else if (expressionTree.is(Tree.Kind.NOT_EQUAL_TO)) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, SymbolicRelation.NOT_EQUAL);
+      } else if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+        evaluateConditionIdentifier(result, executionState, ((IdentifierTree) expressionTree).symbol());
+      } else if (isSuperOrThisMemberSelect(expressionTree)) {
+        evaluateConditionIdentifier(result, executionState, ((MemberSelectExpressionTree) expressionTree).identifier().symbol());
+      } else {
+        for (ExecutionState operatorState : evaluateWithState(executionState, expressionTree)) {
+          if (operatorState.peek() == SymbolicValue.BOOLEAN_FALSE || operatorState.peek() == SymbolicValue.BOOLEAN_TRUE) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)